src/coreclr/jit/importer.cpp

```diff
@@ -4007,6 +4013,24 @@ GenTree* Compiler::impIntrinsic(GenTree*                newobjThis,
                 break;
             }
 
+            case NI_System_Runtime_CompilerServices_RuntimeHelpers_IsKnownConstant:
+            {
+                GenTree* op1 = impPopStack().val;
+                if (op1->OperIsConst() || gtFoldExpr(op1)->OperIsConst())
+                {
+                    // op1 is a known constant, replace with 'true'.
+                    retNode = gtNewIconNode(1);
+                    // We can also consider FTN_ADDR and typeof(T) here
+                }
+                else
+                {
+                    // op1 is not a known constant, replace with 'false'.
+                    // TODO: delay till e.g. morph and try again, maybe op1 will become a const.
+                    retNode = gtNewIconNode(0);
```

Comment:
This drops all side effects of `op1`, it should not do that.

FWIW, in my prototype of the same thing, it really did not take anything to carry this through to morph and fold it there.

Reply:
Ah, right, thanks